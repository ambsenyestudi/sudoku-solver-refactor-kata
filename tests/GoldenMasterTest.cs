using ApprovalTests;
using ApprovalTests.Reporters;
using Moq;
using SudokuSolver.Domain;
using SudokuSolver.Domain.Randomization;
using SudokuSolver.Infrastructure;

namespace SudokuSolver.Test;

[UseReporter(typeof(DiffReporter))]
public class GoldenMasterTest
{
    private readonly Queue<int> randomQueue;
    private StringWriter consoleOutput = new StringWriter();
    private Mock<IRandomizationService> random = new Mock<IRandomizationService>();
    public GoldenMasterTest()
    {
        var file = File.ReadAllText("Random.txt");
        var listRandom = new List<int>();
        foreach (var rawNumber in file.Split("\r\n", StringSplitOptions.RemoveEmptyEntries))
        {
            if(!int.TryParse(rawNumber, out int parsed))
            {
                throw new ArgumentException("Impossible to parse number " + rawNumber);
            }
            listRandom.Add(parsed);
        }
        randomQueue = new Queue<int>(listRandom);
    }
    
    [Fact]
    public void TestReproductibility()
    {
        Console.SetOut(consoleOutput);
        var sudokuSolver = new SudoSolvingService(new OutputService(), random.Object);
        random.Setup(x => x.Next()).Returns(randomQueue.Dequeue);
        random.Setup(x => x.Next(It.IsAny<int>())).Returns(randomQueue.Dequeue);
        sudokuSolver.Play();
        var output = consoleOutput.ToString();
        Approvals.Verify(output);
    }
    

}