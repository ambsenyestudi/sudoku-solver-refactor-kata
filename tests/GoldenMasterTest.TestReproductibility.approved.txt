
Final look of the solved board:
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|259|
|358|792|614|
|912|645|873|
+---+---+---+
|791|423|568|
|536|819|427|
|824|567|391|
+---+---+---+

Starting look of the board to solve:
+---+---+---+
|.7.|...|.36|
|...|..6|7..|
|2.9|.74|.8.|
+---+---+---+
|6..|1.8|..9|
|35.|..2|...|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+

================================================================================

(5, 9) can only contain 4.
+---+---+---+
|.7.|...|.36|
|...|..6|7..|
|2.9|.74|.8.|
+---+---+---+
|6..|1.8|..9|
|35.|..2|..4|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006700209074080600108009350002004012000803000023000530000007800060091

(3, 2) can only contain 6.
+---+---+---+
|.7.|...|.36|
|...|..6|7..|
|269|.74|.8.|
+---+---+---+
|6..|1.8|..9|
|35.|..2|..4|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006700269074080600108009350002004012000803000023000530000007800060091

(4, 2) can only contain 4.
+---+---+---+
|.7.|...|.36|
|...|..6|7..|
|269|.74|.8.|
+---+---+---+
|64.|1.8|..9|
|35.|..2|..4|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006700269074080640108009350002004012000803000023000530000007800060091

(3, 9) can only contain 5.
+---+---+---+
|.7.|...|.36|
|...|..6|7..|
|269|.74|.85|
+---+---+---+
|64.|1.8|..9|
|35.|..2|..4|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006700269074085640108009350002004012000803000023000530000007800060091

(2, 9) can only contain 2.
+---+---+---+
|.7.|...|.36|
|...|..6|7.2|
|269|.74|.85|
+---+---+---+
|64.|1.8|..9|
|35.|..2|..4|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006702269074085640108009350002004012000803000023000530000007800060091

(3, 4) can only contain 3.
+---+---+---+
|.7.|...|.36|
|...|..6|7.2|
|269|374|.85|
+---+---+---+
|64.|1.8|..9|
|35.|..2|..4|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006702269374085640108009350002004012000803000023000530000007800060091

(3, 7) can only contain 1.
+---+---+---+
|.7.|...|.36|
|...|..6|7.2|
|269|374|185|
+---+---+---+
|64.|1.8|..9|
|35.|..2|..4|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006702269374185640108009350002004012000803000023000530000007800060091

(5, 7) can only contain 6.
+---+---+---+
|.7.|...|.36|
|...|..6|7.2|
|269|374|185|
+---+---+---+
|64.|1.8|..9|
|35.|..2|6.4|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006702269374185640108009350002604012000803000023000530000007800060091

(4, 3) can only contain 7.
+---+---+---+
|.7.|...|.36|
|...|..6|7.2|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|35.|..2|6.4|
|.12|...|8.3|
+---+---+---+
|...|.23|...|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006702269374185647108009350002604012000803000023000530000007800060091

(7, 9) can only contain 8.
+---+---+---+
|.7.|...|.36|
|...|..6|7.2|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|35.|..2|6.4|
|.12|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006702269374185647108009350002604012000803000023008530000007800060091

(5, 3) can only contain 8.
+---+---+---+
|.7.|...|.36|
|...|..6|7.2|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|..2|6.4|
|.12|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006702269374185647108009358002604012000803000023008530000007800060091

(5, 5) can only contain 9.
+---+---+---+
|.7.|...|.36|
|...|..6|7.2|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|.92|6.4|
|.12|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|8..|.6.|.91|
+---+---+---+
Code: 070000036000006702269374185647108009358092604012000803000023008530000007800060091

(9, 2) can only contain 2.
+---+---+---+
|.7.|...|.36|
|...|..6|7.2|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|.92|6.4|
|.12|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|82.|.6.|.91|
+---+---+---+
Code: 070000036000006702269374185647108009358092604012000803000023008530000007820060091

(2, 2) can only contain 8.
+---+---+---+
|.7.|...|.36|
|.8.|..6|7.2|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|.92|6.4|
|.12|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|82.|.6.|.91|
+---+---+---+
Code: 070000036080006702269374185647108009358092604012000803000023008530000007820060091

(6, 1) can only contain 9.
+---+---+---+
|.7.|...|.36|
|.8.|..6|7.2|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|.92|6.4|
|912|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|82.|.6.|.91|
+---+---+---+
Code: 070000036080006702269374185647108009358092604912000803000023008530000007820060091

(2, 8) can only contain 4.
+---+---+---+
|.7.|...|.36|
|.8.|..6|742|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|.92|6.4|
|912|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|82.|.6.|.91|
+---+---+---+
Code: 070000036080006742269374185647108009358092604912000803000023008530000007820060091

(1, 7) can only contain 9.
+---+---+---+
|.7.|...|936|
|.8.|..6|742|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|.92|6.4|
|912|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|82.|.6.|.91|
+---+---+---+
Code: 070000936080006742269374185647108009358092604912000803000023008530000007820060091

(5, 4) can only contain 7.
+---+---+---+
|.7.|...|936|
|.8.|..6|742|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|792|6.4|
|912|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|82.|.6.|.91|
+---+---+---+
Code: 070000936080006742269374185647108009358792604912000803000023008530000007820060091

(9, 3) can only contain 4.
+---+---+---+
|.7.|...|936|
|.8.|..6|742|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|792|6.4|
|912|...|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.6.|.91|
+---+---+---+
Code: 070000936080006742269374185647108009358792604912000803000023008530000007824060091

(6, 6) can only contain 5.
+---+---+---+
|.7.|...|936|
|.8.|..6|742|
|269|374|185|
+---+---+---+
|647|1.8|..9|
|358|792|6.4|
|912|..5|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.6.|.91|
+---+---+---+
Code: 070000936080006742269374185647108009358792604912005803000023008530000007824060091

(4, 5) can only contain 3.
+---+---+---+
|.7.|...|936|
|.8.|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|..5|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.6.|.91|
+---+---+---+
Code: 070000936080006742269374185647138009358792604912005803000023008530000007824060091

(6, 5) can only contain 4.
+---+---+---+
|.7.|...|936|
|.8.|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|.45|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.6.|.91|
+---+---+---+
Code: 070000936080006742269374185647138009358792604912045803000023008530000007824060091

(9, 6) can only contain 7.
+---+---+---+
|.7.|...|936|
|.8.|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|.45|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.67|.91|
+---+---+---+
Code: 070000936080006742269374185647138009358792604912045803000023008530000007824067091

(2, 1) can only contain 1.
+---+---+---+
|.7.|...|936|
|18.|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|.45|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.67|.91|
+---+---+---+
Code: 070000936180006742269374185647138009358792604912045803000023008530000007824067091

(1, 3) can only contain 5.
+---+---+---+
|.75|...|936|
|18.|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|.45|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.67|.91|
+---+---+---+
Code: 075000936180006742269374185647138009358792604912045803000023008530000007824067091

(2, 3) can only contain 3.
+---+---+---+
|.75|...|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|.45|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.67|.91|
+---+---+---+
Code: 075000936183006742269374185647138009358792604912045803000023008530000007824067091

(6, 4) can only contain 6.
+---+---+---+
|.75|...|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.67|.91|
+---+---+---+
Code: 075000936183006742269374185647138009358792604912645803000023008530000007824067091

(1, 6) can only contain 1.
+---+---+---+
|.75|..1|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|...|.23|..8|
|53.|...|..7|
|824|.67|.91|
+---+---+---+
Code: 075001936183006742269374185647138009358792604912645803000023008530000007824067091

(7, 2) can only contain 9.
+---+---+---+
|.75|..1|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|.9.|.23|..8|
|53.|...|..7|
|824|.67|.91|
+---+---+---+
Code: 075001936183006742269374185647138009358792604912645803090023008530000007824067091

(9, 4) can only contain 5.
+---+---+---+
|.75|..1|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|.9.|.23|..8|
|53.|...|..7|
|824|567|.91|
+---+---+---+
Code: 075001936183006742269374185647138009358792604912645803090023008530000007824567091

(7, 4) can only contain 4.
+---+---+---+
|.75|..1|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|.9.|423|..8|
|53.|...|..7|
|824|567|.91|
+---+---+---+
Code: 075001936183006742269374185647138009358792604912645803090423008530000007824567091

(1, 5) can only contain 8.
+---+---+---+
|.75|.81|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|.9.|423|..8|
|53.|...|..7|
|824|567|.91|
+---+---+---+
Code: 075081936183006742269374185647138009358792604912645803090423008530000007824567091

(7, 1) can only contain 7.
+---+---+---+
|.75|.81|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|..8|
|53.|...|..7|
|824|567|.91|
+---+---+---+
Code: 075081936183006742269374185647138009358792604912645803790423008530000007824567091

(1, 1) can only contain 4.
+---+---+---+
|475|.81|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|..8|
|53.|...|..7|
|824|567|.91|
+---+---+---+
Code: 475081936183006742269374185647138009358792604912645803790423008530000007824567091

(9, 7) can only contain 3.
+---+---+---+
|475|.81|936|
|183|..6|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|..8|
|53.|...|..7|
|824|567|391|
+---+---+---+
Code: 475081936183006742269374185647138009358792604912645803790423008530000007824567391

(2, 5) can only contain 5.
+---+---+---+
|475|.81|936|
|183|.56|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|..8|
|53.|...|..7|
|824|567|391|
+---+---+---+
Code: 475081936183056742269374185647138009358792604912645803790423008530000007824567391

(8, 5) can only contain 1.
+---+---+---+
|475|.81|936|
|183|.56|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|..8|
|53.|.1.|..7|
|824|567|391|
+---+---+---+
Code: 475081936183056742269374185647138009358792604912645803790423008530010007824567391

(8, 6) can only contain 9.
+---+---+---+
|475|.81|936|
|183|.56|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|..8|
|53.|.19|..7|
|824|567|391|
+---+---+---+
Code: 475081936183056742269374185647138009358792604912645803790423008530019007824567391

(1, 4) can only contain 2.
+---+---+---+
|475|281|936|
|183|.56|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|..8|
|53.|.19|..7|
|824|567|391|
+---+---+---+
Code: 475281936183056742269374185647138009358792604912645803790423008530019007824567391

(2, 4) can only contain 9.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|..8|
|53.|.19|..7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138009358792604912645803790423008530019007824567391

(8, 3) can only contain 6.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|..8|
|536|.19|..7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138009358792604912645803790423008536019007824567391

(7, 7) can only contain 5.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|8.3|
+---+---+---+
|79.|423|5.8|
|536|.19|..7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138009358792604912645803790423508536019007824567391

(6, 8) can only contain 7.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|..9|
|358|792|6.4|
|912|645|873|
+---+---+---+
|79.|423|5.8|
|536|.19|..7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138009358792604912645873790423508536019007824567391

(4, 7) can only contain 2.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|2.9|
|358|792|6.4|
|912|645|873|
+---+---+---+
|79.|423|5.8|
|536|.19|..7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138209358792604912645873790423508536019007824567391

(7, 3) can only contain 1.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|2.9|
|358|792|6.4|
|912|645|873|
+---+---+---+
|791|423|5.8|
|536|.19|..7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138209358792604912645873791423508536019007824567391

(7, 8) can only contain 6.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|2.9|
|358|792|6.4|
|912|645|873|
+---+---+---+
|791|423|568|
|536|.19|..7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138209358792604912645873791423568536019007824567391

(8, 7) can only contain 4.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|2.9|
|358|792|6.4|
|912|645|873|
+---+---+---+
|791|423|568|
|536|.19|4.7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138209358792604912645873791423568536019407824567391

(8, 4) can only contain 8.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|2.9|
|358|792|6.4|
|912|645|873|
+---+---+---+
|791|423|568|
|536|819|4.7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138209358792604912645873791423568536819407824567391

(4, 8) can only contain 5.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|259|
|358|792|6.4|
|912|645|873|
+---+---+---+
|791|423|568|
|536|819|4.7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138259358792604912645873791423568536819407824567391

(5, 8) can only contain 1.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|259|
|358|792|614|
|912|645|873|
+---+---+---+
|791|423|568|
|536|819|4.7|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138259358792614912645873791423568536819407824567391

(8, 8) can only contain 2.
+---+---+---+
|475|281|936|
|183|956|742|
|269|374|185|
+---+---+---+
|647|138|259|
|358|792|614|
|912|645|873|
+---+---+---+
|791|423|568|
|536|819|427|
|824|567|391|
+---+---+---+
Code: 475281936183956742269374185647138259358792614912645873791423568536819427824567391

